src/java.base/share/classes/sun/nio/ch/SocketInputStream.java

```diff
@@ -71,6 +71,18 @@ public int read(byte[] b, int off, int len) throws IOException {
         }
     }
 
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        int timeout = timeoutSupplier.getAsInt();
+        if (!SocketReadEvent.enabled()) {
+            return implRead(b, off, len, timeout);
+        }
+        long start = SocketReadEvent.timestamp();
+        int n = implRead(b, off, len, timeout);
+        SocketReadEvent.offer(start, n, sc.remoteAddress(), timeout);
+        return n;
```

Comment:
This is better, it means that the event is guaranteed to use the same timeout as the read operation.